File: src/analyzer/symbol.js
export const EntityScope = Object.freeze({
  GLOBAL: 'GLOBAL',
  PARAGRAPH: 'PARAGRAPH',
  LOCAL_CLAUSE: 'LOCAL_CLAUSE',
});

export class SymbolEntry {
  constructor(symbolId, name, entityType, firstAppearedNodeId, scope = EntityScope.GLOBAL, isImplicit = false) {
    this.symbol_id = symbolId;
    this.name = name;
    this.entity_type = entityType;
    this.first_appeared_node_id = firstAppearedNodeId;
    this.referenced_node_ids = [firstAppearedNodeId];
    this.scope = scope;
    this.is_implicit = isImplicit;
  }
}

export class SymbolTable {
  constructor() {
    this.symbols = {}; // Keyed by name or unique key
    this.unresolved_references_count = 0;
  }

  add_symbol(name, entityType, nodeId, scope = EntityScope.GLOBAL, isImplicit = false) {
    if (this.symbols[name]) {
      this.symbols[name].referenced_node_ids.push(nodeId);
      return this.symbols[name];
    }
    const symbolId = Object.keys(this.symbols).length;
    const entry = new SymbolEntry(symbolId, name, entityType, nodeId, scope, isImplicit);
    this.symbols[name] = entry;
    return entry;
  }
}

File: tanuki_checker/symbol.py
from typing import Dict, List
from enum import Enum
from pydantic import BaseModel, Field


class EntityScope(str, Enum):
    GLOBAL = "GLOBAL"
    PARAGRAPH = "PARAGRAPH"
    LOCAL_CLAUSE = "LOCAL_CLAUSE"


class SymbolEntry(BaseModel):
    symbol_id: int
    name: str
    entity_type: str  # PER, ORG, LOC, FIRST_PERSON, UNRESOLVED_REF, etc.
    first_appeared_node_id: int
    referenced_node_ids: List[int] = Field(default_factory=list)
    scope: EntityScope = EntityScope.GLOBAL
    is_implicit: bool = False  # 人間文章特有の暗黙の文脈共有フラグ


class SymbolTable(BaseModel):
    symbols: Dict[str, SymbolEntry] = Field(default_factory=dict)
    unresolved_references_count: int = 0

    def add_symbol(
        self,
        name: str,
        entity_type: str,
        node_id: int,
        scope: EntityScope = EntityScope.GLOBAL,
        is_implicit: bool = False,
    ) -> SymbolEntry:
        if name in self.symbols:
            entry = self.symbols[name]
            if node_id not in entry.referenced_node_ids:
                entry.referenced_node_ids.append(node_id)
            return entry
        else:
            symbol_id = len(self.symbols)
            entry = SymbolEntry(
                symbol_id=symbol_id,
                name=name,
                entity_type=entity_type,
                first_appeared_node_id=node_id,
                referenced_node_ids=[node_id],
                scope=scope,
                is_implicit=is_implicit,
            )
            self.symbols[name] = entry
            return entry
